src/s3ql/backends/gdrive.py

```diff
@@ -0,0 +1,345 @@
+
+log = logging.getLogger(__name__)
+
+class Backend(AbstractBackend, metaclass=ABCDocstMeta):
+    '''
+    A backend that stores data in Google Drive.
+    '''
+
+    known_options = set()
+
+    MIME_TYPE_BINARY = 'application/octet-stream'
+    MIME_TYPE_FOLDER = 'application/vnd.google-apps.folder'
+
+    MAX_PROPERTY_SIZE = 124
+    MAX_REF_CHUNK_SIZE = 100
+    VALUE_BASE64 = 'b64:'
+    VALUE_REF = 'ref:'
+
+    def __init__(self, storage_url, login, password, options):
+        '''Initialize local backend
+
+        Login and password are ignored.
+        '''
+        # Unused argument
+        #pylint: disable=W0613
+
+        super().__init__()
+        self.login = login
+        self.password = password
+        self.options = options
+
+        # get google drive service
+        credentials = self._get_credentials()
+        http = credentials.authorize(httplib2.Http())
+        self.service = apiclient.discovery.build('drive', 'v3', http=http)
+
+        # get gdrive folder
+        path = storage_url[len('gdrive://'):].rstrip('/')
+        self.folder = self._lookup_file(path)
+        if self.folder is None or self.folder['mimeType'] != Backend.MIME_TYPE_FOLDER:
+            raise DanglingStorageURLError(self.path)
+
+    def _get_credentials(self):
+        credentials_filename = oauth2client.client._get_environment_variable_file()
+        if not credentials_filename:
+            credentials_filename = oauth2client.client._get_well_known_file()
+        if not credentials_filename:
+            return None
+        with open(credentials_filename) as f:
+            creds = json.load(f)
+
+        credentials = oauth2client.client.SignedJwtAssertionCredentials(
+            creds['client_email'],
+            creds['private_key'],
+            'https://www.googleapis.com/auth/drive',
+            sub=self.login)
+        return credentials
+
+    def _list_files(self, folder, fields):
+        # list folder children
+        page_token = None
+        while True:
+            results = self.service.files().list(
+                q="'{0}' in parents".format(folder['id']),
+                pageToken=page_token,
+                fields="nextPageToken, files(%s)" % fields).execute()
+            for f in results.get('files', []):
+                yield f
+
+            # get next page
+            page_token = results.get('nextPageToken', None)
+            if page_token is None:
+                break
+
+    def _lookup_file(self, name, folder=None):
+        if folder is None:
+            folder = { 'id': 'root', 'path': '/', 'mimeType': Backend.MIME_TYPE_FOLDER }
+
+        # find next path component
+        if len(name) == 0:
+            return folder
+        if name[0] == '/':
+            name = name[1:]
+        n = name.split('/', 1)
+        if len(n[0]) == 0:
+            return folder
+
+        # iterate over folder children
+        for f in self._list_files(folder, "id, name, mimeType, size, properties"):
```

Comment:
Is this really iterating over the contens of a folder, even though we know the exact path? That doesn't seem like a good idea, the folder contents may be huge.